x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/SpatialUtils.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.spatial;
+
+import org.apache.lucene.util.SloppyMath;
+import org.elasticsearch.common.geo.GeoUtils;
+import org.elasticsearch.geometry.Circle;
+import org.elasticsearch.geometry.LinearRing;
+import org.elasticsearch.geometry.Polygon;
+import org.elasticsearch.index.mapper.GeoShapeIndexer;
+
+/**
+ * Utility class for storing different helpful re-usable spatial functions
+ */
+public class SpatialUtils {
+
+    private SpatialUtils() {}
+
+    /**
+     * Makes an n-gon, centered at the provided circle's center, and each vertex approximately
+     * {@link Circle#getRadiusMeters()} away from the center.
+     *
+     * This does not split the polygon across the date-line. Relies on {@link GeoShapeIndexer} to
+     * split prepare polygon for indexing.
+     *
+     * Adapted from from org.apache.lucene.geo.GeoTestUtil
+     * */
+    public static Polygon createRegularGeoShapePolygon(Circle circle, int gons) {
+        double[][] result = new double[2][];
+        result[0] = new double[gons+1];
+        result[1] = new double[gons+1];
+        for(int i=0;i<gons;i++) {
+            double angle = i * (360.0/gons);
+            double x = Math.cos(SloppyMath.toRadians(angle));
+            double y = Math.sin(SloppyMath.toRadians(angle));
+            double factor = 2.0;
+            double step = 1.0;
+            int last = 0;
+
+            // Iterate out along one spoke until we hone in on the point that's nearly exactly radiusMeters from the center:
+            while (true) {
+                double lat = circle.getLat() + y * factor;
+                double lon = circle.getLon() + x * factor;
+                double distanceMeters = SloppyMath.haversinMeters(circle.getLat(), circle.getLon(), lat, lon);
+
+                if (Math.abs(distanceMeters - circle.getRadiusMeters()) < 0.1) {
+                    // Within 10 cm: close enough!
+                    result[0][i] = GeoUtils.normalizeLon(lon);
```

Comment:
I wonder if the problem is more general and the algorithm that splits polygons fails when points consider are coplanar regardless of the proximity. This will probably be a bug in that algorithm, I will check.

Reply:
Indeed, it seems a bug in the way we split polygons. For example this polygon:

```
POLYGON((179 0, -179 0, -178 0, -178 1, 179 1, 179 0))
```

fails with error:

```
org.locationtech.spatial4j.exception.InvalidShapeException: Cannot determine orientation: edges adjacent to (-179.0,0.0) coincide
```